### PR TITLE
Add atomic save to checkpoint routine

### LINKED ISSUE #19970: Checkpoint saving isn't atomic
### Bug description

Checkpoint is no longer atomic. I was implemented [here](https://github.com/Lightning-AI/pytorch-lightning/pull/689) to create checkpoints with ".part" and the rename them, but it is no longer implemented that way looking at code [here](https://github.com/Lightning-AI/pytorch-lightning/blob/f6fd046552a1504023cb3386a8a0df418a810e4f/src/lightning/fabric/utilities/cloud_io.py#L65).

Could we implemented it again in that way? If you kill a job that is training during checkpoint it will corrupted the file

### What version are you seeing the problem on?

master

### How to reproduce the bug

_No response_

### Error messages and logs

```
# Error messages and logs here please
```

### Environment

<details>
  <summary>Current environment</summary>

```
#- Lightning Component (e.g. Trainer, LightningModule, LightningApp, LightningWork, LightningFlow):
#- PyTorch Lightning Version (e.g., 1.5.0):
#- Lightning App Version (e.g., 0.5.2):
#- PyTorch Version (e.g., 2.0):
#- Python version (e.g., 3.9):
#- OS (e.g., Linux):
#- CUDA/cuDNN version:
#- GPU models and configuration:
#- How you installed Lightning(`conda`, `pip`, source):
#- Running environment of LightningApp (e.g. local, cloud):
```

</details>

### More info

_No response_

### PR DESCRIPTION
**Pull request by corwinjoy**
## What does this PR do?
Upgrade the _atomic_checkpoint routine to use a transaction. This should make saves atomic so that if an interrupt happens during a file save, any existing file will not be corrupted.

The following links the related issue to the PR: https://github.com/Lightning-AI/pytorch-lightning/issues/19970

Fixes #19970

I have skipped updating the changelog since it looks like this gets updated as part of the release.

<details>
  <summary><b>Before submitting</b></summary>

- Was this **discussed/agreed** via a GitHub issue? (not for typos and docs)
- [ ] Did you read the [contributor guideline](https://github.com/Lightning-AI/lightning/blob/master/.github/CONTRIBUTING.md), **Pull Request** section?
- [ ] Did you make sure your **PR does only one thing**, instead of bundling different changes together?
- Did you make sure to **update the documentation** with your changes? (if necessary)
- Did you write any **new necessary tests**? (not for typos and docs)
- [ ] Did you verify new and **existing tests pass** locally with your changes?
- Did you list all the **breaking changes** introduced by this pull request?
- Did you **update the CHANGELOG**? (not for typos, docs, test updates, or minor internal changes/refactors)

</details>

## PR review

Anyone in the community is welcome to review the PR.
Before you start reviewing, make sure you have read the [review guidelines](https://github.com/Lightning-AI/lightning/wiki/Review-guidelines). In short, see the following bullet-list:

<details>
  <summary>Reviewer checklist</summary>

- [ ] Is this pull request ready for review? (if not, please submit in draft mode)
- [ ] Check that all items from **Before submitting** are resolved
- [ ] Make sure the title is self-explanatory and the description concisely explains the PR
- [ ] Add labels and milestones (and optionally projects) to the PR so it can be classified

</details>

----
📚 Documentation preview 📚: https://pytorch-lightning--20011.org.readthedocs.build/en/20011/

### CODE REVIEW

File: src/lightning/fabric/utilities/cloud_io.py [hunk 1]
```diff
@@ -76,7 +76,10 @@ def _atomic_save(checkpoint: Dict[str, Any], filepath: Union[str, Path]) -> None
     bytesbuffer = io.BytesIO()
     log.debug(f"Saving checkpoint: {filepath}")
     torch.save(checkpoint, bytesbuffer)
-    with fsspec.open(filepath, "wb") as f:
+
+    # We use a transaction here so that if the save gets interrupted, existing checkpoints are not trashed.
+    fs, _ = fsspec.core.url_to_fs(str(filepath))
```
Comment (awaelchli) [inline src/lightning/fabric/utilities/cloud_io.py:81]:
```suggestion
    fs = get_filesystem(filepath)
```

File: src/lightning/fabric/utilities/cloud_io.py [hunk 2]
```diff
@@ -76,7 +76,10 @@ def _atomic_save(checkpoint: Dict[str, Any], filepath: Union[str, Path]) -> None
     bytesbuffer = io.BytesIO()
     log.debug(f"Saving checkpoint: {filepath}")
     torch.save(checkpoint, bytesbuffer)
-    with fsspec.open(filepath, "wb") as f:
+
+    # We use a transaction here so that if the save gets interrupted, existing checkpoints are not trashed.
```
Comment (awaelchli) [inline src/lightning/fabric/utilities/cloud_io.py:80]: The comment should go above the line it corresponds to :)

Comment (awaelchli) [inline src/lightning/fabric/utilities/cloud_io.py:80]:
```suggestion
    # We use a transaction here to avoid file corruption if the save gets interrupted
```
"existing checkpoints are not trashed." would contradict what transaction means.

**Review by awaelchli — COMMENTED** [review]
Looks great. Just a few suggestions

File: src/lightning/fabric/utilities/cloud_io.py (hunk 1 above)
Comment (corwinjoy) [reply src/lightning/fabric/utilities/cloud_io.py]: That's a reasonable suggestion. I acutally just committed a change to do this a bit differently and use the returned urlpath in conjunction with the filesystem since I think that's a bit more idiomatic with fsspec.

File: src/lightning/fabric/utilities/cloud_io.py (hunk 2 above)
Comment (corwinjoy) [reply src/lightning/fabric/utilities/cloud_io.py]: Applied change.

**Review by awaelchli — APPROVED** [review]

**Comment by awaelchli** [comment]
Thanks @corwinjoy 
I also made a follow-up issue for the distributed checkpoint saving logic.

**Comment by corwinjoy** [comment]
Thanks for the help @awaelchli !

**Comment by brandon-smith-xperi** [comment]
Hi, this PR changes the permissions of the saved checkpoints on macOS and Linux (not sure about Windows).  I suspect that was unintentional, and if so, should a new ticket be filed?

On macOS:
Previous permissions: 644 (this is the expected permissions for a new file)
New permissions: 600 (this is not the expected permissions for a new file)

This snippet shows previous vs new behavior.

```
import io
import fsspec
import torch
from fsspec.core import url_to_fs

x = torch.zeros(1)
bytesbuffer = io.BytesIO()
torch.save(x, bytesbuffer)

# original correct permissions
with fsspec.open('correct_permissions.pt', "wb") as f:
    f.write(bytesbuffer.getvalue())

# new incorrect permissions
fs, urlpath = fsspec.core.url_to_fs('incorrect_permissions.pt')
with fs.transaction, fs.open(urlpath, "wb") as f:
    f.write(bytesbuffer.getvalue())
```

Thanks

**Comment by corwinjoy** [comment]
Yes, this change to the default file permissions is unintended. This seems
like a problem with fsspec, though, I don't see why they would use
nonstandard file permissions there. I think a first step might be to open a
ticket there and understand the rationale behind it. Unless you already
have a reference on why they are doing this?

**Comment by brandon-smith-xperi** [comment]
Thanks @corwinjoy, I'll file a ticket with fsspec
